zb-db/src/main/java/io/camunda/zeebe/db/impl/rocksdb/transaction/ForeignKeyChecker.java

```diff
@@ -30,31 +31,62 @@ public ForeignKeyChecker(
     enabled = settings.enableForeignKeyChecks();
   }
 
-  private void assertExists(
-      final ZeebeTransaction transaction, final DbForeignKey<? extends DbKey> foreignKey)
+  public void assertExists(
+      final ZeebeTransaction transaction, final ContainsForeignKeys containsForeignKey)
       throws Exception {
+    if (!enabled) {
+      return;
+    }
+    for (final var fk : containsForeignKey.containedForeignKeys()) {
+      assertForeignKeyExists(transaction, fk);
+    }
+  }
+
+  private void assertForeignKeyExists(
+      final ZeebeTransaction transaction, final DbForeignKey<DbKey> foreignKey) throws Exception {
+    if (foreignKey.skip().test(foreignKey.inner())) {
+      return;
+    }
+
     keyBuffer.putLong(0, foreignKey.columnFamily().ordinal(), ZeebeDbConstants.ZB_DB_BYTE_ORDER);
     foreignKey.write(keyBuffer, Long.BYTES);
-    final var value =
-        transaction.get(
-            transactionDb.getDefaultNativeHandle(),
-            transactionDb.getReadOptionsNativeHandle(),
-            keyBuffer.byteArray(),
-            Long.BYTES + foreignKey.getLength());
-    if (value == null) {
+    final var exists =
+        switch (foreignKey.match()) {
+          case Full -> keyExists(transaction, foreignKey);
+          case Prefix -> prefixExists(transaction, foreignKey);
+        };
+    if (!exists) {
       throw new ZeebeDbInconsistentException(
           "Foreign key " + foreignKey.inner() + " does not exist in " + foreignKey.columnFamily());
     }
   }
 
-  public void assertExists(
-      final ZeebeTransaction transaction, final ContainsForeignKeys containsForeignKey)
+  private boolean keyExists(
+      final ZeebeTransaction transaction, final DbForeignKey<? extends DbKey> foreignKey)
       throws Exception {
-    if (!enabled) {
-      return;
-    }
-    for (final var fk : containsForeignKey.containedForeignKeys()) {
-      assertExists(transaction, fk);
+    return transaction.get(
+            transactionDb.getDefaultNativeHandle(),
+            transactionDb.getReadOptionsNativeHandle(),
+            keyBuffer.byteArray(),
+            Long.BYTES + foreignKey.getLength())
+        != null;
+  }
+
+  private boolean prefixExists(
+      final ZeebeTransaction transaction, final DbForeignKey<? extends DbKey> foreignKey) {
+    try (final var iterator =
+        transaction.newIterator(
+            transactionDb.getPrefixReadOptions(), transactionDb.getDefaultHandle())) {
+      final var prefixKey = keyBuffer.byteArray();
```

Comment:
:wrench: both methods (this and `keyExists`) use `prefixKey` and `prefixLength`. So you could pass in those instead of`foreignKey. Passing in `foreignKey`is a bit strange, because it is pretty much not really used, because it was already written into `keyBuffer` beforehand. Personally, I think passing in `keyBuffer.byteArray()` and `length` would make it clearer what these methods do. It also would avoid confusion if somebody else wants to call these methods and forgot to update the `keyBuffer` beforehand.

Reply:
I like your suggestion but to throw custom exceptions in both `assertPrefixExists` and `assertKeyExists` as suggested here: https://github.com/camunda/zeebe/pull/8960#discussion_r832599898 requires that we pass in the `foreignKey` so that we can set a helpful exception message.